### DISCUSSION TITLE
Custom context with superRefine

### DISCUSSION THREAD

**Discussion by oof2win2** [colinhacks/zod#1087] [Q&A]
Hey there, is there a possibility of adding a custom context (say with a `this` or some other parameter) to a .superRefine addition? I need to use an external variable that I cannot get during the superRefine run, so I need to have something like this:
```ts
const allIDs = [1, 3, 6]
const schema = z.object({
  id: z.number()
})
  .superRefine((data, ctx) => {
    if (!allIDs.includes(data.id)) ctx.addIssue...
})
```
Would it however be possible to have something like this instead? It allows for the functions to be pure and be ran in cases where you obtain the data *some way* (such as by a database query or an API call) to allow for the schema to have only validation code, rather than also including code to fetch the data itself.
```ts
const schema = z.object({
  id: z.number()
})
  .superRefine((data, ctx, params) => {
    const {allIDs} = params
    if (!allIDs.includes(data.id)) ctx.addIssue...
})
```

**Comment by JohannesKlauss** [comment]
I was also wondering if this is possible. Injecting contextual runtime data (without fetching it all the time from the server via `.refine(async val => ...)`) seems like a nice feature. Is there something like this planned? I'd like to help out implementing it.

**Comment by tacomanator** [comment]
It would be great if this could be done even higher up at the chain, at the parse level: `schema.parse(data, context)`, such that we could use the context in transform or overwrite as well as refinements, but without it ending up in the result data. In my particular case, I am using zod schemas to add derived properties to objects. I can simply add a `_context` property, but that creates other issues/limitations.
